pkg/rtc/participant.go

```diff
@@ -666,50 +668,84 @@ func (p *ParticipantImpl) GetAudioLevel() (level uint8, active bool) {
 	return
 }
 
-func (p *ParticipantImpl) GetConnectionQuality() livekit.ConnectionQuality {
+func (p *ParticipantImpl) GetConnectionQuality() *livekit.ConnectionQualityInfo {
 	// avg loss across all tracks, weigh published the same as subscribed
 	var pubLoss, subLoss uint32
 	var reducedQualityPub bool
 	var reducedQualitySub bool
+	var audioScores float64
+	var numAudioTracks int
 	p.lock.RLock()
 	defer p.lock.RUnlock()
 	for _, pubTrack := range p.publishedTracks {
 		if pubTrack.IsMuted() {
 			continue
 		}
-		pubLoss += pubTrack.PublishLossPercentage()
-		publishing, registered := pubTrack.NumUpTracks()
-		if registered > 0 && publishing != registered {
-			reducedQualityPub = true
+		// audio has scores available
+		if pubTrack.Kind() == livekit.TrackType_AUDIO {
+			audioScores += pubTrack.GetConnectionScore()
+			numAudioTracks++
+		} else {
+			pubLoss += pubTrack.PublishLossPercentage()
+			publishing, registered := pubTrack.NumUpTracks()
+			if registered > 0 && publishing != registered {
+				reducedQualityPub = true
+			}
 		}
 	}
-	numTracks := uint32(len(p.publishedTracks))
-	if numTracks > 0 {
-		pubLoss /= numTracks
-	}
 
 	for _, subTrack := range p.subscribedTracks {
 		if subTrack.IsMuted() {
 			continue
 		}
-		if subTrack.DownTrack().GetForwardingStatus() != sfu.ForwardingStatusOptimal {
-			reducedQualitySub = true
+		if subTrack.DownTrack().Kind() == webrtc.RTPCodecTypeAudio {
+			audioScores += subTrack.DownTrack().GetConnectionScore()
+			numAudioTracks++
+		} else {
+			if subTrack.DownTrack().GetForwardingStatus() != sfu.ForwardingStatusOptimal {
```

Comment:
Forwarder isn't computing this value correctly for connection quality needs. today if the publisher is only publishing two layers, but subscriber wants highest quality, we would actually penalize the subscriber quality score as not getting the optimal stream.

I believe `GetForwardingStatus` should be modified to return `Optimal` as long as the subscriber is getting the highest *available* layer.

Reply:
Yeah, this requires more work.

Currently, if `Forwarder` is pausing, it is reporting `off` which is incorrect. I will look at correcting that in `Forwarder` code.

Reply:
Made a PR - https://github.com/livekit/livekit-server/pull/267 to fix this.